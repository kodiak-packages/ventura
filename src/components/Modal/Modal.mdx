---
title: Modal
---

import { Playground, Props } from 'docz';
import { Button, Modal, Input } from '../..';

# Modal

A modal is a dialog that appears on top of the main content to require user interaction. This modal disables the main content until the user explicitly interacts with it.

## Best practices

- Wrapping your buttons inside `Modal.Footer` will allow them to be automatically displayed right

## Examples

### Basic usage

<Playground>
  {() => {
    const [isOpen, setIsOpen] = React.useState(false);
    return (
      <>
        <Button onClick={() => setIsOpen(true)}>Open basic modal</Button>
        <Modal title="This is the title" isOpen={isOpen}>
          <span>Insert your text or form elements here</span>
          <Modal.Footer>
            <Button type='secondary' onClick={() => setIsOpen(false)}>Cancel</Button>
            <Button onClick={() => 'do some logic here'}>Create</Button>
          </Modal.Footer>
        </Modal>
      </>
    );
}}

</Playground>

### Modal with async action

<Playground>
  {() => {
    const [isOpen, setIsOpen] = React.useState(false);
    const [isLoading, setIsLoading] = React.useState(false);
    const createHandler = async () => {
      // await a promise here
      setIsLoading(true);
      setTimeout(() => {setIsLoading(false); setIsOpen(false)}, 1000);
    }
    return (
      <>
        <Button onClick={() => setIsOpen(true)}>Create Todo</Button>
        <Modal title="Create a Todo" isOpen={isOpen}>
          <Input />
          <Modal.Footer>
            <Button type='secondary' onClick={() => setIsOpen(false)}>Cancel</Button>
            <Button onClick={createHandler} isLoading={isLoading}>Create</Button>
          </Modal.Footer>
        </Modal>
      </>
    );
}}

</Playground>

## API

### Modal

<Props of={Modal} />

### Modal.Footer

<Props of={Modal.Footer} />
