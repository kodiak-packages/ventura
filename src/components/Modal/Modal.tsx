import React, { ReactNode, useEffect } from 'react';
import { Portal } from 'react-portal';
import classNames from 'classnames';

import Footer from './ModalFooter/ModalFooter';

import cssReset from '../../css-reset.module.css';
import styles from './Modal.module.css';

type Props = {
  title?: string;
  isOpen?: boolean;
  onEscKeyDown?: () => void;
  className?: string;
  children?: ReactNode;
};

const disableBodyScroll = (isDisabled: boolean) => {
  if (isDisabled) {
    document.body.style.overflow = 'hidden';
  } else {
    document.body.style.overflow = 'inherit';
  }
};

const Modal: React.FC<Props> & { Footer: typeof Footer } = ({
  title,
  isOpen = false,
  onEscKeyDown,
  className,
  children,
}: Props) => {
  const modalRef = React.useRef<HTMLDivElement>(null);

  useEffect(() => {
    disableBodyScroll(isOpen);
    // eslint-disable-next-line no-unused-expressions
    if (isOpen) modalRef.current?.focus();
  }, [isOpen, modalRef]);

  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if (event.keyCode === 27) {
      if (onEscKeyDown) onEscKeyDown();
    }
  };

  const overlayClassNames = classNames(cssReset.ventura, styles.overlay, className);

  return isOpen ? (
    <Portal>
      <div className={overlayClassNames} role="article" data-testid="modal">
        <article
          ref={modalRef}
          className={styles.modal}
          onKeyDown={handleKeyDown}
          // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
          tabIndex={0}
          role="presentation"
        >
          {Boolean(title) && <h1 className={styles.title}>{title}</h1>}
          {children}
        </article>
      </div>
    </Portal>
  ) : null;
};

Modal.Footer = Footer;

export default Modal;
