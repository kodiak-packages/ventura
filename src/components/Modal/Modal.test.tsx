import React from 'react';
import { fireEvent, getQueriesForElement, render } from '@testing-library/react';

import Modal from './Modal';

describe('Modal', () => {
  test('snapshot when shown', () => {
    // portals don't render in the fragment so we have look in baseElement
    // https://github.com/testing-library/react-testing-library/issues/62#issuecomment-513199414
    const { baseElement, rerender } = render(
      <Modal title="This is the title" isOpen>
        <span>Insert your text or form elements here</span>
        <Modal.Footer>Buttons are placed here</Modal.Footer>
      </Modal>,
    );
    expect(baseElement).toMatchSnapshot();

    rerender(
      <Modal title="This is the title" isOpen maxWidth="800px">
        <span>Insert your text or form elements here</span>
        <Modal.Footer>Buttons are placed here</Modal.Footer>
      </Modal>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  test('snapshot when hidden', () => {
    const { baseElement } = render(<Modal>Invisible</Modal>);
    const modal = getQueriesForElement(baseElement).queryByTestId('modal');
    expect(modal).toBe(null);
  });

  test('className prop', () => {
    const className = 'center';
    const { baseElement } = render(
      <Modal isOpen className={className}>
        Content
      </Modal>,
    );
    const overlayElement = getQueriesForElement(baseElement).getByText(/Content/).parentElement!;

    const renderedClassNames = overlayElement.className.split(' ');
    expect(renderedClassNames).toContain(className);
    // className in prop should be the last in the row
    expect(renderedClassNames.indexOf(className)).toBe(renderedClassNames.length - 1);
  });

  test('modal should close on pressing ESC', () => {
    const mockOnEscKeyDown = jest.fn();
    const { baseElement } = render(
      <Modal isOpen onEscKeyDown={mockOnEscKeyDown}>
        Content
      </Modal>,
    );
    const modalElement = getQueriesForElement(baseElement).getByText(/Content/)!;
    fireEvent.keyDown(modalElement, { key: 'Escape', code: 'Escape', keyCode: 27 });
    expect(mockOnEscKeyDown).toHaveBeenCalledTimes(1);
  });

  test('the modal should receive focus on open', () => {
    const { baseElement } = render(<Modal isOpen>Content</Modal>);
    const modalElement = getQueriesForElement(baseElement).getByText(/Content/)!;
    expect(document.activeElement).toBe(modalElement);
  });
});
