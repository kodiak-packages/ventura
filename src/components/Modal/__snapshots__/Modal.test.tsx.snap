// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal snapshot when shown 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div>
    <div
      class="ventura overlay"
      data-testid="modal"
      role="article"
    >
      <article
        class="modal"
        role="presentation"
        style="max-width: 650px;"
        tabindex="0"
      >
        <h1
          class="title"
        >
          This is the title
        </h1>
        <span>
          Insert your text or form elements here
        </span>
        <div
          class="container"
        >
          Buttons are placed here
        </div>
      </article>
    </div>
  </div>
</body>
`;

exports[`Modal snapshot when shown 2`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div>
    <div
      class="ventura overlay"
      data-testid="modal"
      role="article"
    >
      <article
        class="modal"
        role="presentation"
        style="max-width: 800px;"
        tabindex="0"
      >
        <h1
          class="title"
        >
          This is the title
        </h1>
        <span>
          Insert your text or form elements here
        </span>
        <div
          class="container"
        >
          Buttons are placed here
        </div>
      </article>
    </div>
  </div>
</body>
`;
